### PR TITLE
Add standalone ASM billing support

### PR DESCRIPTION
**Pull request by jandro996**
# What Does This Do

Add new boolean environment variable `DD_EXPERIMENTAL_APPSEC_STANDALONE_ENABLED`, when it's enabled:

- Libraries must add the numeric tag `_dd.apm.enabled:0` to the metrics map of the service entry spans. `_dd.apm.enabled` is assumed to be 1 when absent, so it is backward compatible.
- Disable the generation of APM trace metrics by disabling the computation of the APM trace metrics and the computation agent-side of the APM trace metrics by pretending it was already done by the library (the trace payload sent to the agent must contain the `HTTP header Datadog-Client-Computed-Stats: yes`)
- Introduce a new propagated span tag `_dd.p.appsec: 1` providing the knowledge to downstream services that the current distributed trace is containing at least one ASM event and must inherit from the given force-keep priority indeed. 
- Ignore the force-keep priority in the absence of this propagated `_dd.p.appsec` span tag
- Use a new `TimeSampler` to only allow 1 APM trace per minute  as standalone ASM is only interested in the traces containing ASM events. But the service catalog and the billing need a continuous ingestion of at least at 1 trace per minute to consider a service as being live and billable. In the absence of ASM events, no APM traces must be sent, so we need to let some regular APM traces go through, even in the absence of ASM events.
- If ASM standalone billing is enabled and here is no ASM events (No _dd.p.appsec) propagation should be stopped to downstream services 

# Motivation

ASM is a natural continuation of APM, leveraging concepts such as traces to build threat monitoring protection capabilities, or on telemetry to build vulnerability management.
Though, some customers (primarily infrastructure-monitoring-only customers) only want ASM. We want to make this possible, still internally leveraging APM and provide the same service to ASM customers, while allowing them to not use APM

image

# Additional Notes

Jira ticket: [APPSEC-10459]

[APPSEC-10459]: https://datadoghq.atlassian.net/browse/APPSEC-10459?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: internal-api/src/main/java/datadog/trace/api/Config.java [hunk 1]
```diff
@@ -4060,6 +4068,10 @@ public Boolean getAppSecScaEnabled() {
     return appSecScaEnabled;
   }
 
+  public boolean isExperimentalAppSecStandaloneEnabled() {
+    return experimentalAppSecStandaloneEnabled;
+  }
```
Comment (ValentinZakharov) [inline internal-api/src/main/java/datadog/trace/api/Config.java:4073]: I think, better keep naming like `isAppSecStandaloneEnabled` (without word `experimental`) for methods, but keep it for option string `experimental.appsec.standalone.enabled`.
In this way, when option will be turned to non-experimental, you will not need to do a renaming of the methods.

Comment (jandro996) [reply internal-api/src/main/java/datadog/trace/api/Config.java:4073]: The name is defined by the RFC so we can't change it.

File: dd-java-agent/agent-iast/src/main/java/com/datadog/iast/Reporter.java [hunk 2]
```diff
@@ -99,6 +99,7 @@ private VulnerabilityBatch getOrCreateVulnerabilityBatch(final AgentSpan span) {
       // TODO: We need to check if we can have an API with more fine-grained semantics on why traces
       // are kept.
       segment.setTagTop(DDTags.MANUAL_KEEP, true);
+      segment.setTagTop(Tags.PROPAGATED_APPSEC, true);
```
[the remark this answers is not in the document]
Comment (ValentinZakharov) [reply dd-java-agent/agent-iast/src/main/java/com/datadog/iast/Reporter.java:102]: based on PR description, value should be numeric value

**Comment by ValentinZakharov** [comment]
I don't, get why for `_dd.apm.enabled` and `_dd.p.appsec` are used numeric values, but not booleans?
is there any other possible values except `0` and `1` ?

File: dd-java-agent/appsec/src/main/java/com/datadog/appsec/gateway/AppSecRequestContext.java [hunk 3]
```diff
@@ -415,6 +419,15 @@ public void reportEvents(Collection<AppSecEvent> events) {
       }
       try {
         this.collectedEvents.addAll(events);
+        // If APM is disabled Provide the knowledge to downstream services that the current
+        // distributed trace is containing at least one ASM event and must inherit from the
+        // given force-keep priority indeed
```
Comment (ValentinZakharov) [inline dd-java-agent/appsec/src/main/java/com/datadog/appsec/gateway/AppSecRequestContext.java:424]:
```suggestion
        // If APM is disabled, inform downstream services that the current
        // distributed trace contains at least one ASM event and must inherit
        // the given force-keep priority
```

File: dd-java-agent/appsec/src/main/java/com/datadog/appsec/gateway/AppSecRequestContext.java [hunk 4]
```diff
@@ -415,6 +419,15 @@ public void reportEvents(Collection<AppSecEvent> events) {
...
+        if (Config.get().isExperimentalAppSecStandaloneEnabled()) {
+          AgentSpan activeSpan = AgentTracer.activeSpan();
+          if (activeSpan != null) {
+            activeSpan.getLocalRootSpan().setTag(Tags.PROPAGATED_APPSEC, true);
+          }
+        }
```
Comment (ValentinZakharov) [inline dd-java-agent/appsec/src/main/java/com/datadog/appsec/gateway/AppSecRequestContext.java:430]: I'm sorry, but I think, better place setting propagate appsec to `GatewayBrige` in `requestEnded` callback, where all appSec events are stored to local root span

File: dd-java-agent/agent-iast/src/main/java/com/datadog/iast/Reporter.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (jandro996) [reply dd-java-agent/agent-iast/src/main/java/com/datadog/iast/Reporter.java:102]: The propagation header should be 1 but I used the same approach of the manual keep and use a boolean, in the end this tag is intercepted to set the propagated one. The final root span has a _dd.p.appsec -> "1" and not a 'p.appsec' -> true

File: dd-java-agent/agent-iast/src/test/groovy/com/datadog/iast/ReporterTest.groovy [hunk 5]
```diff
@@ -80,6 +80,7 @@ class ReporterTest extends DDSpecification {
       ]
     }''', batch.toString(), true)
     1 * traceSegment.setTagTop('manual.keep', true)
+    1 * traceSegment.setTagTop('p.appsec', true)
```
[the remark this answers is not in the document]
Comment (jandro996) [reply dd-java-agent/agent-iast/src/test/groovy/com/datadog/iast/ReporterTest.groovy:83]: As I've commented before the tag is intercepted and is not present in the root span

File: dd-smoke-tests/asm-standalone-billing/src/test/groovy/datadog/smoketest/asmstandalonebilling/AbstractAsmStandaloneBillingSmokeTest.groovy [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    return trace.spans[0].metrics['_sampling_priority_v1'] == priority
+  }
+
+  protected isSampledBySampler(DecodedTrace trace) {
+    def samplingPriority = trace.spans[0].metrics['_sampling_priority_v1']
+    return samplingPriority == PrioritySampling.SAMPLER_KEEP || samplingPriority == PrioritySampling.SAMPLER_DROP
+  }
+
+  protected hasAppsecPropagationTag(DecodedTrace trace) {
+    return trace.spans[0].meta['_dd.p.appsec'] == "1"
```
[the remark this answers is not in the document]
Comment (jandro996) [reply dd-smoke-tests/asm-standalone-billing/src/test/groovy/datadog/smoketest/asmstandalonebilling/AbstractAsmStandaloneBillingSmokeTest.groovy:62]: yes 👍

File: dd-smoke-tests/asm-standalone-billing/src/test/groovy/datadog/smoketest/asmstandalonebilling/AsmStandaloneBillingConerCasesSmokeTest.groovy [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+      .get().build()
+
+    when:
+    final response1 = client.newCall(request).execute()
+
+    then: "First trace should be sampled with sampler keep as it's the first span checked in a minute"
+    response1.successful
+    waitForTraceCount(1)
+    checkRootSpanPrioritySampling(traces[0], PrioritySampling.USER_KEEP)
+
```
[the remark this answers is not in the document]
Comment (jandro996) [reply dd-smoke-tests/asm-standalone-billing/src/test/groovy/datadog/smoketest/asmstandalonebilling/AsmStandaloneBillingConerCasesSmokeTest.groovy:91]: TBH I'm thinking on remove this corner case test as they are not related to ASM Stand alone billing and we habe many tests

File: dd-java-agent/appsec/src/main/java/com/datadog/appsec/gateway/AppSecRequestContext.java (hunk 4 above)
Comment (jandro996) [reply dd-java-agent/appsec/src/main/java/com/datadog/appsec/gateway/AppSecRequestContext.java:430]: If I understood the code correctly, the main problem with set the propagation tag in `requestEnded` is that the downstream service has been called before, so the downstream service will never receive  the propagation tag

**Comment by jandro996** [comment]
> I don't, get why for `_dd.apm.enabled` and `_dd.p.appsec` are used numeric values, but not booleans? is there any other possible values except `0` and `1` ?

Is an RFC specification, you can check it [here](https://docs.google.com/document/d/12NBx-nD-IoQEMiCRnJXneq4Be7cbtSc6pJLOFUWTpNE/edit?pli=1)

File: dd-trace-core/src/test/groovy/datadog/trace/core/propagation/DatadogPropagationTagsTest.groovy [hunk 8]
```diff
@@ -134,6 +136,29 @@ class DatadogPropagationTagsTest extends DDCoreSpecification {
...
+    propagationTags.headerValue(PropagationTags.HeaderType.DATADOG) == expectedHeaderValue
+    propagationTags.createTagMap() == tags
+
+    where:
+    originalTagSet                         | enabled       | expectedHeaderValue   | tags
+    // keep the existing dm tag as is
+    "_dd.p.appsec=1"                       | true          | "_dd.p.appsec=1"      | ["_dd.p.appsec": "1"]
+    ""                                     | false         | null                  | [:]
+    //Invalid input
+    "_dd.p.appsec=0"                       | false         | null                  | ["_dd.propagation_error": "decoding_error"]
```
[the remark this answers is not in the document]
Comment (jandro996) [reply dd-trace-core/src/test/groovy/datadog/trace/core/propagation/DatadogPropagationTagsTest.groovy:159]: I can change it, I think it was defined in that way in the previous versions of the RFC but seems that right now we can open it to 0 or 1

File: internal-api/src/main/java/datadog/trace/api/Config.java (hunk 1 above)
Comment (jandro996) [reply internal-api/src/main/java/datadog/trace/api/Config.java:4073]: My fault, I misundertood @ValentinZakharov comment, I'll change it

**Comment by ValentinZakharov** [comment]
Can we introduce one more `Config.isApmEnabled()` to be controversial to `isExperimentalAppSecStandaloneEnabled()`?
In many places of the code I can see something like:
```
// If APM is disabled...
if (Config.get().isAppSecStandaloneEnabled()) { ... }
```
so, the method `!Config.isApmEnabled()` will be much clearer than `isAppSecStandaloneEnabled`

File: dd-java-agent/appsec/src/main/java/com/datadog/appsec/gateway/AppSecRequestContext.java [hunk 9]
```diff
@@ -415,6 +419,15 @@ public void reportEvents(Collection<AppSecEvent> events) {
       }
       try {
         this.collectedEvents.addAll(events);
+        // If APM is disabled, inform downstream services that the current
+        // distributed trace contains at least one ASM event and must inherit
+        // the given force-keep priority
+        if (Config.get().isAppSecStandaloneEnabled()) {
```
Comment (ValentinZakharov) [inline dd-java-agent/appsec/src/main/java/com/datadog/appsec/gateway/AppSecRequestContext.java:425]: I think counter-method like `Config.isApmEnabled()` will bring more clearance to code + will let us get rid of some comments.

File: dd-trace-core/src/main/java/datadog/trace/common/metrics/MetricsAggregatorFactory.java [hunk 10]
```diff
@@ -10,7 +10,11 @@ public class MetricsAggregatorFactory {
 
   public static MetricsAggregator createMetricsAggregator(
       Config config, SharedCommunicationObjects sharedCommunicationObjects) {
-    if (config.isTracerMetricsEnabled()) {
+    if (config.isTracerMetricsEnabled()
+        && !config.isAppSecStandaloneEnabled() // When APM is disabled, libraries must
```
Comment (ValentinZakharov) [inline dd-trace-core/src/main/java/datadog/trace/common/metrics/MetricsAggregatorFactory.java:14]: Same, method `Config.isApmEnabled()` looks better here

File: dd-trace-core/src/main/java/datadog/trace/core/PendingTrace.java [hunk 11]
```diff
@@ -488,12 +489,16 @@ public void setSamplingPriorityIfNecessary() {
...
+    if (traceConfig.sampler instanceof PrioritySampler && rootSpan != null) {
+      // Ignore the force-keep priority in the absence of propagated _dd.p.appsec span tag.
+      if (Config.get().isAppSecStandaloneEnabled()
+          && !rootSpan.context().getPropagationTags().isAppsecPropagationEnabled()) {
+        ((PrioritySampler) traceConfig.sampler).setSamplingPriority(rootSpan);
+        return;
+      }
+      if (rootSpan.context().getSamplingPriority() == PrioritySampling.UNSET) {
+        ((PrioritySampler) traceConfig.sampler).setSamplingPriority(rootSpan);
+      }
```
Comment (ValentinZakharov) [inline dd-trace-core/src/main/java/datadog/trace/core/PendingTrace.java:501]: I'm sorry, I didn't get, does both `if-blocks` are doing exactly the same?

File: dd-java-agent/appsec/src/main/java/com/datadog/appsec/gateway/AppSecRequestContext.java (hunk 4 above)
Comment (jandro996) [reply dd-java-agent/appsec/src/main/java/com/datadog/appsec/gateway/AppSecRequestContext.java]: yes, I've changed it

File: dd-trace-core/src/main/java/datadog/trace/core/PendingTrace.java (hunk 11 above)
Comment (jandro996) [reply dd-trace-core/src/main/java/datadog/trace/core/PendingTrace.java:501]: I will do it more readable, thanks!

File: dd-trace-core/src/main/java/datadog/trace/common/sampling/AsmStandaloneSampler.java [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+public class AsmStandaloneSampler implements Sampler, PrioritySampler {
+
+  private static final Logger log = LoggerFactory.getLogger(AsmStandaloneSampler.class);
+
+  private final AtomicLong lastSampleTime;
+  private final int rateInMilliseconds;
+
+  public AsmStandaloneSampler(int rateInMilliseconds) {
+    this.rateInMilliseconds = rateInMilliseconds;
+    this.lastSampleTime = new AtomicLong(-1);
```
Comment (manuel-alvarez-alvarez) [inline dd-trace-core/src/main/java/datadog/trace/common/sampling/AsmStandaloneSampler.java:19]:
```suggestion
    this.lastSampleTime = new AtomicLong(System.currentTimeMillis() - rateInMilliseconds);
```

File: dd-trace-core/src/main/java/datadog/trace/common/sampling/AsmStandaloneSampler.java [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+  private boolean shouldSample() {
+    long now = System.currentTimeMillis();
+    long lastTime = lastSampleTime.get();
+
+    if (lastTime == -1 || now - lastTime >= rateInMilliseconds) {
+      if (lastSampleTime.compareAndSet(lastTime, now)) {
+        return true;
+      }
+    }
+    return false;
```
Comment (manuel-alvarez-alvarez) [inline dd-trace-core/src/main/java/datadog/trace/common/sampling/AsmStandaloneSampler.java:50]:
```suggestion
    return lastSampleTime.updateAndGet(value -> now - value >= rateInMilliseconds ? now : value) == now;
```

File: dd-java-agent/appsec/src/main/java/com/datadog/appsec/gateway/AppSecRequestContext.java (hunk 9 above)
Comment (jandro996) [reply dd-java-agent/appsec/src/main/java/com/datadog/appsec/gateway/AppSecRequestContext.java]: I agree with you, but this will be done in the next iteration (there is a plan to decouple APM from tracing). I've been talking about this with @smola and seems that for this PR we are fine highlighting the stand alone mode with the variable

File: dd-java-agent/appsec/src/main/java/com/datadog/appsec/powerwaf/PowerWAFModule.java [hunk 14]
```diff
@@ -444,6 +456,21 @@ public void onDataAvailable(
           }
         }
         Collection<AppSecEvent> events = buildEvents(resultWithData);
+        // If APM is disabled, inform downstream services that the current
+        // distributed trace contains at least one ASM event and must inherit
+        // the given force-keep priority
+        if (Config.get().isAppSecStandaloneEnabled()) {
+          if (rateLimiter == null || !rateLimiter.isThrottled()) {
```
Comment (jandro996) [inline dd-java-agent/appsec/src/main/java/com/datadog/appsec/powerwaf/PowerWAFModule.java:463]: @ValentinZakharov, @cataphract could you check if this change related with keep appsec rate limiter decision for standalone biling?

**Review by PerfectSlayer — COMMENTED** [review]
Hi @jandro996 

First, thanks for the detailed PR description! It helped a lot to get into the PR changes.
Then, sorry for the delay 🙏  (a lot of stuff to review for Dash 😮‍💨 )

The overall change looks great. I left few comments:
* One about config for metric deactivation (not clear where the check should be done)
* Few about design to try to move product logic away from core (at least the best we could considering the current architecture),
* Few about value caching, documentation and code coverage
* And some leftovers (just to make sure they are in a state your think it's right to merge them now)

File: dd-trace-core/src/main/java/datadog/trace/core/PendingTrace.java [hunk 15]
```diff
@@ -488,12 +489,13 @@ public void setSamplingPriorityIfNecessary() {
...
-
-    if (traceConfig.sampler instanceof PrioritySampler
-        && rootSpan != null
-        && rootSpan.context().getSamplingPriority() == PrioritySampling.UNSET) {
-
-      ((PrioritySampler) traceConfig.sampler).setSamplingPriority(rootSpan);
+    if (traceConfig.sampler instanceof PrioritySampler && rootSpan != null) {
+      // Ignore the force-keep priority in the absence of propagated _dd.p.appsec span tag.
+      if ((Config.get().isAppSecStandaloneEnabled()
+              && !rootSpan.context().getPropagationTags().isAppsecPropagationEnabled())
```
[the remark this answers is not in the document]
Comment (jandro996) [reply dd-trace-core/src/main/java/datadog/trace/core/PendingTrace.java:495]: If I understand correctly what you are proposing, I don't think it can be done that way :S , as we need to be able to sample a trace with any sampling priority and  the original code only reach the sampler for `PrioritySampling.UNSET`. 
For instance if the user sets a manual keep, with your suggestion the APPSEC specific check will never be reached.

File: dd-trace-core/src/main/java/datadog/trace/core/DDSpanContext.java [hunk 16]
```diff
@@ -532,6 +537,11 @@ private boolean setThisSpanSamplingPriority(final int newPriority, final int new
     if (!validateSamplingPriority(newPriority, newMechanism)) {
       return false;
     }
+    if (Config.get().isAppSecStandaloneEnabled() && newMechanism == SamplingMechanism.APPSEC) {
+      SAMPLING_PRIORITY_UPDATER.set(this, newPriority);
+      propagationTags.updateTraceSamplingPriority(newPriority, SamplingMechanism.APPSEC);
+      return true;
+    }
```
[the remark this answers is not in the document]
Comment (jandro996) [reply dd-trace-core/src/main/java/datadog/trace/core/DDSpanContext.java:549]: I have tried to solve it in a slightly different way :sweat_smile:, let me know if it works for you

File: dd-trace-core/src/main/java/datadog/trace/common/writer/ddagent/DDAgentApi.java [hunk 17]
```diff
@@ -108,7 +109,10 @@ public Response sendSerializedTraces(final Payload payload) {
               .addHeader(DATADOG_DROPPED_SPAN_COUNT, Long.toString(payload.droppedSpans()))
               .addHeader(
                   DATADOG_CLIENT_COMPUTED_STATS,
-                  metricsEnabled && featuresDiscovery.supportsMetrics() ? "true" : "")
+                  (metricsEnabled && featuresDiscovery.supportsMetrics())
+                          || Config.get().isAppSecStandaloneEnabled()
```
[the remark this answers is not in the document]
Comment (jandro996) [reply dd-trace-core/src/main/java/datadog/trace/common/writer/ddagent/DDAgentApi.java:113]: We need this as the RFC says that we need to cheat the agent with this header:

"Disabling the computation agent-side of the APM trace metrics by pretending it was already done by the library.
To do so, the trace payload sent to the agent must contain the HTTP header Datadog-Client-Computed-Stats: yes"

Long story short, when asm standalone billing is enabled, APM metrics should be disabled but `HTTP header Datadog-Client-Computed-Stats: yes` should be added to the request

File: dd-trace-core/src/main/java/datadog/trace/common/metrics/MetricsAggregatorFactory.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (jandro996) [reply dd-trace-core/src/main/java/datadog/trace/common/metrics/MetricsAggregatorFactory.java:14]: To be honest we decided to add the condition at this point to avoid problems in other pieces of code and highlight the points where the ASM standalone mode modifies the current behaviour to facilitate the future refactor.

We need to disable the APM metrics but sending the ` HTTP header Datadog-Client-Computed-Stats: yes` to the agent

File: dd-trace-core/src/main/java/datadog/trace/core/propagation/CorePropagation.java [hunk 18]
```diff
@@ -58,6 +59,15 @@ private <C> void inject(
...
 
+    /**
+     * If the experimental appsec standalone feature is enabled and appsec propagation is disabled
+     * (no ASM events), stop propagation
+     */
+    if (Config.get().isAppSecStandaloneEnabled()
+        && !ddSpanContext.getPropagationTags().isAppsecPropagationEnabled()) {
+      return;
+    }
+
```
[the remark this answers is not in the document]
Comment (jandro996) [reply dd-trace-core/src/main/java/datadog/trace/core/propagation/CorePropagation.java:70]: I'm not entirely sure I understand your comment 😅. Are you saying there's a better place to perform this check and not send the context, or that this will be done in a future implementation? The main issue is that we need to have this feature available for a client by the end of the quarter.

**Comment by jandro996** [comment]
> Hi @jandro996
> 
> First, thanks for the detailed PR description! It helped a lot to get into the PR changes. Then, sorry for the delay 🙏 (a lot of stuff to review for Dash 😮‍💨 )
> 
> The overall change looks great. I left few comments:
> 
> * One about config for metric deactivation (not clear where the check should be done)
> * Few about design to try to move product logic away from core (at least the best we could considering the current architecture),
> * Few about value caching, documentation and code coverage
> * And some leftovers (just to make sure they are in a state your think it's right to merge them now)

Thanks a lot for your review! 

I've made some changes based on your suggestions, although I couldn't fully implement all of them as you suggested. Additionally, regarding your comments about APM metrics, the main idea is:

**Disable the APM metrics** but sending the  **HTTP header Datadog-Client-Computed-Stats: yes** to the agent

File: dd-trace-core/src/main/java/datadog/trace/core/propagation/CorePropagation.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (PerfectSlayer) [reply dd-trace-core/src/main/java/datadog/trace/core/propagation/CorePropagation.java:70]: This will be done later, as part of some refactoring. I was just pointing some place that needs more love from the platform team in term of design improvements.
There is not much more you can do for now 😉

File: dd-trace-core/src/main/java/datadog/trace/core/PendingTrace.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (PerfectSlayer) [reply dd-trace-core/src/main/java/datadog/trace/core/PendingTrace.java:495]: You're right, you can't... (I misread the `||` part).
I don't like the idea of adding product checks within tracing but I struggle to find another solution until we get sampling fixed for APPSEC (unified sampling flag / decision that includes APPSEC only specific sampling rules).

File: dd-trace-core/src/main/java/datadog/trace/core/DDSpanContext.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (PerfectSlayer) [reply dd-trace-core/src/main/java/datadog/trace/core/DDSpanContext.java]: I was fine putting the logic inside the existing `validateWithSamplingPriority()` method, especially since this part 
```
|| priority == SAMPLER_KEEP; // Necessary for ASM standalone billing
```
will only be needed when `Config.get().isAppSecStandaloneEnabled()` is true, right?

File: dd-trace-core/src/main/java/datadog/trace/common/metrics/MetricsAggregatorFactory.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (jandro996) [reply dd-trace-core/src/main/java/datadog/trace/common/metrics/MetricsAggregatorFactory.java]: I've changed this as you suggested me, as turn off the metrics in standalone billing is a expected behaviour 👍

[the remark this answers is not in the document]
Comment (jandro996) [reply dd-trace-core/src/main/java/datadog/trace/common/metrics/MetricsAggregatorFactory.java]: I've changed this as you suggested me, as turn off the metrics in standalone billing is a expected behaviour 👍

**Review by PerfectSlayer — APPROVED** [review]
Don’t forget to squash your commit when/before merging 😉
